src/typecheck/eq.rs

```diff
@@ -43,24 +43,61 @@
 //! times. For anything more complex, we return false.
 
 use super::*;
-use crate::eval;
-use crate::term::UnaryOp;
+use crate::{eval, term::UnaryOp};
 
 /// The maximal number of variable links we want to unfold before abandoning the check. It should
 /// stay low, but has been fixed arbitrarily: feel fee to increase reasonably if it turns out
 /// legitimate type equalities between simple contracts are unduly rejected in practice.
 pub const MAX_GAS: u8 = 8;
 
+/// Abstract over the term environment, which is represented differently in the typechecker and
+/// during evaluation.
+///
+/// The evaluation environment holds [`crate::eval::lazy::Thunks`], which are `Rc<RefCell<_>>`
+/// under the hood, while the term environment used during typechecking is just maps identifiers to
+/// a pair `(RichTerm, Environment)`. To have an interface that works with both,
+/// `TermEnvironment::get` has to take a closure representing the continuation of the task to do
+/// with the result instead of merely returning it.
+pub trait TermEnvironment: Clone {
```

Comment:
Does it really need to be it's own trait and multiple reimplementations? It looks like a method on (the very invented syntax) `GenericEnvironment<Ident, V> where V: Into<(Richterm, Self)>` or something like that.

Reply:
I think it would'nt be trivial, because in the case of the eval env, you have thunks, which are not `Into<(RichTerm, Self)>`. The code of `get_then` is quite different for both cases, so my first reaction is I doubt you can do only one generic implementation. If something, you may offset the differences to another trait like `BorrowableAs<(RichTerm, Env)>` and have `get_then` in the environment, but is that really better, if in any case we need a trait which is implemented only by `SimpleTermEnvironment` and  `eval::Environment`?

Also, this need is currently quite specific to contract equality checking. When you only handle one unique shape of environment, which is what's happening usually, you don't need this abstraction. Would that be sensible to keep it like this for now, and to think about moving it into `GenericEnvironment` if one day it is required elsewhere?

Reply:
Fair enough, that holds.